File: C++/684-Redundant_Connection.cpp
// Runtime: 8 ms, faster than 78.88% of C++ online submissions for Redundant Connection.
// Memory Usage: 9.9 MB, less than 76.47% of C++ online submissions for Redundant Connection.
class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        vector<int> root(edges.size() + 1, -1);
        for (vector<int> edge : edges) {
            int x = find(root, edge[0]);
            int y = find(root, edge[1]);
            if (x == y)
                return edge;
            root[x] = y;
        }
        vector<int> res = {-1, -1};
        return res;
    }

    int find(const vector<int>& root, int x) {
        while (root[x] != -1)
            x = root[x];
        return x;
    }
};

File: Python3.x/684-Redundant_Connection.py
# union-find
# Runtime: 68 ms, faster than 59.57% of Python3 online submissions for Redundant Connection.
# Memory Usage: 14.2 MB, less than 14.29% of Python3 online submissions for Redundant Connection.
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        self.root = [-1] * (len(edges) + 1)
        for edge in edges:
            x = self.find(edge[0])
            y = self.find(edge[1])
            if x == y:
                return edge
            self.root[x] = y
        return [-1, -1]

    def find(self, x):
        while self.root[x] != -1:
            x = self.root[x]
        return x
